examples/tree_source/tree_source/app.py

```diff
@@ -1,119 +1,150 @@
-    def __init__(self, decade):
-        self.decade = decade
-        self._data = []
-
-    # Display values for the decade in the tree.
-    @property
-    def year(self):
-        return "{}0's".format(self.decade)
+
+class LoadingFailedNode:
+    """A node to represent failed loading of children"""
+
+    def __init__(self, parent):
+        self.parent = parent
+        self.children = []
+        self.name = 'loading failed'
+        self.date_modified = ''
 
     # Methods required for the data source interface
     def __len__(self):
-        return len(self._data)
+        return 0
 
     def __getitem__(self, index):
-        return self._data[index]
+        pass
+
+    @staticmethod
+    def can_have_children():
+        return False
 
-    def can_have_children(self):
-        return True
 
+class Node:
+    """A node which loads its children on-demand."""
 
-class DecadeSource(Source):
-    def __init__(self):
+    def __init__(self, path, parent):
         super().__init__()
-        self._decades = []
 
+        self.parent = parent
+        self._children = []
+
+        self.path = path
+        self._mtime = os.stat(self.path).st_mtime
+        self._icon = toga.Icon('resources/folder.icns')
+
+        self._did_start_loading = False
+
+    # Methods required for the data source interface
     def __len__(self):
-        return len(self._decades)
+        return len(self.children)
 
     def __getitem__(self, index):
-        return self._decades[index]
+        return self.children[index]
 
-    def add(self, entry):
-        decade = entry['year'] // 10
+    def can_have_children(self):
+        # this will trigger loading of children, if not yet done
+        return len(self.children) > 0
+
+    # Property that returns the first column value as (icon, label)
+    @property
+    def name(self):
+        return self._icon, os.path.basename(self.path)
+
+    # Property that returns modified date as str
+    @property
+    def date_modified(self):
+        return datetime.fromtimestamp(self._mtime).strftime("%d %b %Y at %H:%M")
+
+    # on-demand loading of children
+    @property
+    def children(self):
+        if not self._did_start_loading:
+            self._did_start_loading = True
+            self.load_children()
+        return self._children
+
+    def load_children(self):
         try:
-            decade_root = {
-                root.decade: root
-                for root in self._decades
-            }[decade]
-        except KeyError:
-            decade_root = Decade(decade)
-            self._decades.append(decade_root)
-            self._decades.sort(key=lambda v: v.decade)
-        movie = Movie(**entry)
-        decade_root._data.append(movie)
-        self._notify('insert', parent=decade_root, index=len(decade_root._data) - 1, item=movie)
+            names = os.scandir(self.path)
+            sub_paths = [os.path.join(self.path, p) for p in names if os.path.isdir(os.path.join(self.path, p))]
```

Comment:
As a "modern Python style" thing, it would be preferable to use pathlib, rather than os.join.

Reply:
Also - note to self: This demo currently breaks on GTK - scandir returns DirEntry objects, not strings. This might be addressed using pathlib iteration; but we need to make sure the demo still works on GTK.

Reply:
I have switched generally from os.path to pathlib now, it does result in much cleaner code.